### PR TITLE
Basic token validation implementation

### PR DESCRIPTION
**Pull request by jo-arroyo**
Basic implementation of token validation for the `@azure/node-token-validation` library, to replace `passport-azure-ad`. 

This PR includes:
- Node-token-validation library with basic implementation of token validation: 
  - JWT parsing
  - JWT type check
  - JWT signature check
  - JWT claims check
- A simple Node sample that calls the `validateToken` API (other public APIs will be added in another PR)

Due to the length of this PR, and given that implementation is incremental, the following will be included in separate PRs:
- Unit tests
- Additional public APIs
- Updated response type
- Updated metadata refresh handling
- Documentation

More detail can be found in the initial design spec. I will be writing additional design specs to address: 
- Configuration and TokenValidationParameters
- Public APIs
- Metadata and resiliency

### CODE REVIEW

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/api/TokenValidator.ts]: Can these sorts log be trace logs?

Comment (jasonnutter) [reply lib/node-token-validation/src/api/TokenValidator.ts]: And I might have it be something like:

```suggestion
        this.logger.trace("TokenValidator.validateToken called");
```

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/api/TokenValidator.ts]: What is missing here?

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/api/TokenValidator.ts]: To be clear, this is not an error condition?

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/api/TokenValidator.ts]: Why only half? (would be good to put answer to that in the comment as well)

Appreciate teh comments!

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/config/OpenIdConfigProvider.ts]: Is that try/catch needed if just rethrowing?

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/config/OpenIdConfigProvider.ts]: v2.0 is just for AAD. We may want to have that be added based on protocol mode (like we do for browser).

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/config/TokenValidationParameters.ts]: One other nuance is to accont for trailing slash on the `api://{client-id}/` format.

Comment (jasonnutter) [reply lib/node-token-validation/src/config/TokenValidationParameters.ts]: I'm curious what other libraries do here...they may require an application to provide these explicitly.

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (jasonnutter) [inline lib/node-token-validation/src/config/TokenValidationParameters.ts]: I think we'll want to spread both objects here, instead of checking truthiness for each property, as that would potentially improperly use the default value if the provided value is false/0/null.

```suggestion
    const overlaidParams: ValidationParameters = {
        ...DEFAULT_VALIDATION_PARAMS,
        ...DEFAULT_ID_TOKEN_PARAMS,
        ...params
    };
```

**Review by jasonnutter — CHANGES_REQUESTED** [review]
Looking good! Left a few comments. Also, please add the following:

- updated CODEOWNERS
- Fill in README

Edit: Ah, missed that you will have incremental PRs on top of this. Feel free to include these updates there.

**Comment by jasonnutter** [comment]
Would also be good to get the label bot updated. @tnorling

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts]: Yeah, that's an improvement. Changed.

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/OpenIdConfigProvider.ts]: Good point. Removed.

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/OpenIdConfigProvider.ts]: Right. Added protocolMode option.

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/TokenValidationParameters.ts]: After some investigation, it seems that other libraries require providing these explicitly, which seems better. Removed these defaults.

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/TokenValidationParameters.ts]: Thanks for the catch. Changed.

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts]: Ah. I was figuring out whether type validation was needed. Have since amended.

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts]: Upon further examination, this would be better as an error condition. Thanks for the catch!

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts]: This is how the OIDC spec specifies `c_hash` and `at_hash` should be validated. See [this link](https://openid.net/specs/openid-connect-core-1_0.html#HybridTokenValidation) at 3.3.2.11.

Added this link to the comments for clarity!

File: .github/workflows/build-test-pr.yml
Comment (tnorling) [inline .github/workflows/build-test-pr.yml]: Did we make this a separate job because it doesn't have "msal" in the name? If so, I almost prefer updating the `lib-build-test` job to be more generic so we don't have to duplicate the logic

File: change/@azure-node-token-validation-7f710590-0a83-448e-9a03-d398955429f8.json
Comment (tnorling) [inline change/@azure-node-token-validation-7f710590-0a83-448e-9a03-d398955429f8.json]:
```suggestion
  "comment": "Implement basic token validation #4368",
```

File: lib/node-token-validation/package.json
Comment (tnorling) [inline lib/node-token-validation/package.json]: Should this be 12+?

File: lib/node-token-validation/package.json
Comment (tnorling) [inline lib/node-token-validation/package.json]: I think this was not updated in the other libraries when we moved to Jest.
```suggestion
    "clean:coverage": "rimraf ./coverage/*",
```

File: samples/node-token-validation-samples/basic-sample/package-lock.json
Comment (tnorling) [inline samples/node-token-validation-samples/basic-sample/package-lock.json]: Should we remove this?

File: lib/node-token-validation/package.json
Comment (jasonnutter) [reply lib/node-token-validation/package.json]: I would say even versions above 12:

```suggestion
    "node": "12 || 14 || 16"
```

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (tnorling) [inline lib/node-token-validation/src/api/TokenValidator.ts]: Why is this async? Is this something that will be added onto later?

**Comment by tnorling** [comment]
There's a couple other workflows/scripts/configs that will need to (eventually) be updated (e.g. labeler, audit, release, issue templates and bot, lerna, typedoc). I'll let you decide if you want to do that as part of this PR or a follow up PR

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (pkanher617) [reply lib/node-token-validation/src/api/TokenValidator.ts]: +1, not sure why these functions are async or return promises

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (pkanher617) [inline lib/node-token-validation/src/config/TokenValidationParameters.ts]: Can we add some comments here detailing the difference between `TokenValidationParameters` and `ValidationParameters`?

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (pkanher617) [inline lib/node-token-validation/src/config/TokenValidationParameters.ts]: Is it possible to use `Pick` or `Omit` and have `TokenValidationParameters` and `ValidationParameters` share properties? Maintaining these separately might be add on complexity.

File: .github/workflows/build-test-pr.yml
Comment (jo-arroyo) [reply .github/workflows/build-test-pr.yml]: Makes sense. Updated.

File: samples/node-token-validation-samples/basic-sample/package-lock.json
Comment (jo-arroyo) [reply samples/node-token-validation-samples/basic-sample/package-lock.json]: Yes! Removed.

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts]: Ah, this was left over from when we were going to use custom validators. Changed these back to sync. Thanks for the catch.

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/TokenValidationParameters.ts]: Yes, added.

File: lib/node-token-validation/src/config/TokenValidationParameters.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/TokenValidationParameters.ts]: That's a fair point. I used Partial to add it for now, but if there ends up being more parameters needed I may separate it out to different files.

File: lib/node-token-validation/src/crypto/CryptoProvider.ts
Comment (tnorling) [inline lib/node-token-validation/src/crypto/CryptoProvider.ts]: Please address the lint warning here

File: samples/node-token-validation-samples/basic-sample/README.md
Comment (tnorling) [inline samples/node-token-validation-samples/basic-sample/README.md]: Do we want to link to the msal-browser samples here?

File: samples/node-token-validation-samples/basic-sample/README.md
Comment (tnorling) [inline samples/node-token-validation-samples/basic-sample/README.md]: Some of this doesn't match the rest of the PR. Seems like this README was written for a previous iteration?

File: samples/node-token-validation-samples/basic-sample/package.json
Comment (tnorling) [inline samples/node-token-validation-samples/basic-sample/package.json]:
```suggestion
    "build:package": "cd ../../../lib/msal-common && npm run build && cd ../node-token-validation && npm run build",
```

File: samples/node-token-validation-samples/basic-sample/README.md
Comment (jo-arroyo) [reply samples/node-token-validation-samples/basic-sample/README.md]: Thanks for the catch. Updated this!

File: samples/node-token-validation-samples/basic-sample/README.md
Comment (jo-arroyo) [reply samples/node-token-validation-samples/basic-sample/README.md]: Good idea. Linked.

File: change/@azure-node-token-validation-7f710590-0a83-448e-9a03-d398955429f8.json
Comment (tnorling) [inline change/@azure-node-token-validation-7f710590-0a83-448e-9a03-d398955429f8.json]: Should this be none?
```suggestion
  "type": "none",
```

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (tnorling) [inline lib/node-token-validation/src/api/TokenValidator.ts]: Can we add some typedoc comments to these APIs?

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (tnorling) [inline lib/node-token-validation/src/config/OpenIdConfigProvider.ts:42]: Should we use the `Authority` class exported by msal-common here instead, since it already implements this logic?

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts]: Sure. Added typedoc comments.

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/config/OpenIdConfigProvider.ts:42]: I explored using the `Authority` class from msal-common, but decided against it for the time being because it requires a `cacheManager` implementation, along with other `AuthorityOptions` which may not be relevant to this library. In the interest of keeping the first version of this library lightweight, I have not yet implemented caching, but may do so to address metadata refresh concerns later on, at which time it may make sense to use the `Authority` class instead.

**Review by tnorling — APPROVED** [review]

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (sameerag) [inline lib/node-token-validation/src/api/TokenValidator.ts:161]: Should we also document the specific claims being validated here in comments? or is this covered in `BaseValidationParameters` definition?

File: lib/node-token-validation/src/config/OpenIdConfigProvider.ts
Comment (sameerag) [inline lib/node-token-validation/src/config/OpenIdConfigProvider.ts:42]: Same comment as @tnorling, lets evaluate `Authority` class to be leveraged at some point.

File: lib/node-token-validation/src/network/HttpClient.ts
Comment (sameerag) [inline lib/node-token-validation/src/network/HttpClient.ts:61]: Can we consider `node-fetch`? Also `axios` default implementation does not support proxy out of the box, please check #4447 for details.

**Review by sameerag — APPROVED** [review]
Approved with some minor comments!

File: lib/node-token-validation/src/network/HttpClient.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/network/HttpClient.ts:61]: Offline discussion determined that we would keep `fetch` from node in consideration, but would not immediately upgrade. Support for proxies will come in another PR.

File: lib/node-token-validation/src/api/TokenValidator.ts
Comment (jo-arroyo) [reply lib/node-token-validation/src/api/TokenValidator.ts:161]: Updated comments to be more clear about specific claims. Additional claims may be added as public APIs are added.

**Review by jasonnutter — APPROVED** [review]

**Comment by jasonnutter** [comment]
Might be good to squash these commits when merging (should be an option from teh dropdown). Thanks for addressing feedback, great job so far!

**Pull request merged**
